packages/web3-core-types/src/Hex/Hex.js

```diff
@@ -0,0 +1,378 @@
+/*
+    This file is part of web3.js.
+
+    web3.js is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    web3.js is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @file Hex.js
+ * @author Oscar Fonseca <hiro@cehh.io>
+ * @date 2019
+ */
+
+import {isNumber, isString} from 'lodash';
+import utf8 from 'utf8';
+
+export default class Hex {
+    /**
+     * @dev Wrap as object
+     * @param {String} hex
+     *
+     * @constructor
+     */
+    constructor(params) {
+        const requires = ['hex'];
+
+        this.props = {};
+
+        if (isString(params)) {
+            params = {
+                hex: params
+            };
+        }
+
+        if (Hex.isValid(params.hex)) {
+            this.props.hex = params.hex.toString();
+        }
+
+        if (params.hex === 'empty') {
+            this.props.hex = '0x';
+        }
+
+        requires.forEach((propName) => {
+            if (typeof this.props[propName] === 'undefined') {
+                this._throw(propName, params[propName]);
+            }
+        });
+
+        /* Make the props immutable */
+        Object.freeze(this.props);
```

Comment:
We should wrap this object with an es6 proxy and with a strict getter and setter handler.

Reply:
Since `this.props` are meant to be the immutable, shouldn't it be enough with getters? The core idea was to use `props` are the internal database, and any modifications to the props would require creating a new object. Is this too strict?